Add item renderer with model transformation mode, deprecate old one.

i509VCB opened the pull request:
This PR simply adds a new interface which supplies the additional `Mode` parameter.
All old renderers are delegated to an implementation of the new renderer, thereby making this PR still backwards compatible with all existing renderers.

i509VCB committed:
Add item renderer with model transformation mode, deprecate old one.

This PR simply adds a new interface which supplies the additional `Mode` parameter.
All old renderers are delegated to an implementation of the new renderer, thereby making this PR still backwards compatable with all existing renderers.

(cherry picked from commit 7c9162e99f48c6e3989eec9a46afaa935d65ce30)

i509VCB commented:
Resolves #972

Juuxel reviewed (changes requested):
The name of the new interface is ugly. I still think a better way would be just to add the parameter, since 1.16.2 is a good opportunity for breaking changes.

i509VCB commented:
The parameter exists in .1

I dont want to name it `BuiltinItemRenderer2`

i509VCB commented:
Also knowing how these PRs go, 1.16.2 will be out by then and changing the Param name will break stuff

liach commented:
@i509VCB Just add a new default method in the interface which default implementation logs a warning for the users to implement this new default method than the old method without transformation mode. Then remove that old method in like 1.16.2. I agree with juuz that your class name choice cannot be uglier.

i509VCB commented:
My issue with the log thing is you'll be hitting the logger potentially upwards of a thousands a second with some horrible item renderers.

i509VCB commented:
I'd be fine compromising on the name if no other real alternative exists.

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BuiltinItemRendererWithMode.java:
```diff
@@ -0,0 +1,43 @@
+	 * @param mode            the model transformation mode
+	 * @param matrices        the matrix stack
+	 * @param vertexConsumers the vertex consumer provider
+	 * @param light           the color light multiplier at the rendering position
```
I do not think this is a multiplier.

```suggestion
	 * @param light           packed lightmap coordinates
```

i509VCB committed:
Update fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BuiltinItemRendererWithMode.java

Co-authored-by: shartte <shartte@users.noreply.github.com>

i509VCB commented:
Talked to player a bit, he suggested `DynamicItemRenderer` as a name for the new interface

i509VCB committed:
Rename new interfact to `DynamicItemRenderer`, make it an inner class of the registry class.

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/BuiltinItemRendererRegistryImpl.java:
```diff
@@ -50,7 +59,23 @@ public void register(Item item, BuiltinItemRenderer renderer) {
+		private final BuiltinItemRenderer renderer;
+
+		private OldRendererImpl(BuiltinItemRenderer renderer) {
+			this.renderer = renderer;
```
can just convert this to a lambda imo since that is not performance-wise worse

i509VCB replied:
Just messy as a Lambda imo. But I could

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/BuiltinItemRendererRegistryImpl.java:
```diff
@@ -32,13 +36,18 @@
 
 	@Override
 	public void register(Item item, BuiltinItemRenderer renderer) {
+		this.register(item, new OldRendererImpl(Objects.requireNonNull(renderer, "renderer is null")));
```
```suggestion
		Objects.requireNonNull(renderer, "renderer is null")
		this.register(item, (stack, mode, matrices, vertexConsumers, light, overlay) -> renderer.render(stack, matrices, vertexConsumers, light, overlay));
```
And remove that old renderer class

i509VCB committed:
Update fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/BuiltinItemRendererRegistryImpl.java

Co-authored-by: liach <7806504+liach@users.noreply.github.com>

i509VCB committed:
Update BuiltinItemRendererRegistryImpl.java

i509VCB committed:
Update BuiltinItemRendererRegistryImpl.java

i509VCB committed:
Imports

i509VCB committed:
Use ItemConvertible for new render method

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BuiltinItemRendererRegistry.java:
```diff
@@ -42,6 +47,43 @@
+	 */
+	@FunctionalInterface
+	@Environment(EnvType.CLIENT)
+	interface DynamicItemRenderer {
```
I think this would be better as a top-level type like the old interface.

liach replied:
Given this name, it's better left at top level indeed. :+1:

i509VCB committed:
Rename new interfact to `DynamicItemRenderer`, make it an inner class of the registry class.

fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BuiltinItemRendererRegistry.java

i509VCB committed:
Move inner class out, fix formatting issue in loot-tables with linux oses

i509VCB committed:
why was this multilined

i509VCB commented on fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/FabricLootSupplierBuilder.java:
```diff
@@ -81,7 +81,7 @@ public FabricLootSupplierBuilder copyFrom(LootTable supplier) {
 
 	/**
-	 * Copies the pools and functions of the {@code supplier} to this builder.
+	 * Copies the pools and functions of the {@code supplier} to this builder.
```
So a story on this one, apparently a symbol was there that prevents gradle from building on my arch setup, so the whitespace had to change

Juuxel replied:
Ah, I've probably made an accidental NBSP there (iirc that's my code?)

modmuss50 replied:
Can checksyle fail on these? idk

i509VCB committed:
license header lol

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/BuiltinItemRendererRegistryImpl.java:
```diff
@@ -56,8 +51,21 @@ public void register(ItemConvertible item, BuiltinItemRenderer renderer) {
+		Objects.requireNonNull(item.asItem(), "item is null");
+		Objects.requireNonNull(renderer, "renderer is null");
+
+		if (RENDERERS.containsKey(item.asItem())) {
```
```suggestion
		if (RENDERERS.put(item.asItem(), renderer) != null) {
```

Juuxel replied:
This will still overwrite the existing entry if the modder decides to catch any exceptions produced by this method, which is incorrect behaviour IMO. If that exception is thrown, there should be no side effects to the registry itself.

liach replied:
then `putIfAbsent`

i509VCB replied:
Will do that in a bit

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/BuiltinItemRendererRegistryImpl.java:
```diff
@@ -56,8 +51,21 @@ public void register(ItemConvertible item, BuiltinItemRenderer renderer) {
+		if (RENDERERS.containsKey(item.asItem())) {
+			throw new IllegalArgumentException("Item " + Registry.ITEM.getId(item.asItem()) + " already has a builtin renderer!");
+		}
+
+		RENDERERS.put(item.asItem(), renderer);
```
```suggestion
```

i509VCB committed:
Apply suggestions from code review

Co-authored-by: liach <7806504+liach@users.noreply.github.com>

Player3324 reviewed (approved):
Looks good, thanks! One detail to consider is moving the interface to a nested class to avoid spamming otherwise useless classes.

Juuxel reviewed (changes requested):
See https://github.com/FabricMC/fabric/pull/973#discussion_r474060856

modmuss50 commented:
Ready for last call once that last minor issue is sorted.

i509VCB commented:
Funny enough player I had the nested class originally but of course complaints galore

i509VCB committed:
Make renderer a nested class again

NinjaPhenix reviewed (approved):
Looks good to me and thanks for not breaking existing apis.

i509VCB committed:
putIfAbsent
